csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+      pad_tv->getDataType().value());
+  IrBuilder::create<PadOp>(new_out, pad_tv, merged_pad_widths, pad_value);
+  return new_out;
+}
+
+// Note [ PadOp Propagation Rule ]
+//
+// Push PadOp to its producer to reduce segmentation caused by the resize node
+// introduced in PadOp. The hope is that we can either push PadOp to inputs to
+// the fusion, or far enough that the fusion segments before the PadOp would
+// become a trivial no-op.
+//
+// The concept is that the following program:
+//   tv1 = UnaryOp(tv0)
+//   tv2 = PadOp(tv1)
+// can be replayed as the program below:
+//   tv0_padded = PadOp(tv0)
+//   tv2_new = UnaryOp(tv0_padded)
+// Given certain constraints on the UnaryOp, we can ensure that the
+// computational output remains the same when we replace all uses of `tv2` with
+// `tv2_new`.
+//
+// This function only propagates `zero-padding`, i.e. PadOp with pad value as
+// constant zero. Based on which we have certain operations that can allow PadOp
+// to propagated across. See `padCompatibleUnaryOp` and `padCompatibleBinaryOp`.
+Val* propagatePadToProducer(PadOp* pad_op) {
+  // establish pad dependencies, ensures that pad_value and pad_widths are live
+  // at the time of replay.
+  std::vector<Val*> pad_dependencies;
+  for (Val* val : pad_op->inputs()) {
+    if (val == pad_op->in() || val->isConst()) {
+      continue;
+    }
+    pad_dependencies.push_back(val);
+  }
+  auto pad_replay_check = [&pad_dependencies](Val* val) {
+    if (!val->isA<TensorView>()) {
+      return false;
+    }
+    if (std::any_of(
+            pad_dependencies.begin(),
+            pad_dependencies.end(),
+            [val](Val* pad_dependency) {
+              return DependencyCheck::isDependencyOf(pad_dependency, val);
+            })) {
+      return false;
+    }
+    return true;
+  };
+
+  // NOTE: We skip the propagation when any of the three conditions is true:
+  //   1. The optimization logic assumes a zero pad. This is used for logic in
+  //   handling binary operations, see Note [ PadOp Propagation Rule] ; or 2.a
+  //   if `pad_op->in()` is used more than by the pad_op; or 2.b if
+  //   `pad_op->in()` is an output tv.
+  // for case 2.a/b, because pad_op->in() needs to stay alive after the
+  // propagation, we want to avoid adding another data flow branch.
+  if (!pad_op->value()->isZero() || pad_op->in()->uses().size() > 1 ||
+      pad_op->in()->isFusionOutput()) {
+    return nullptr;
+  }
+
+  // frontier contains `Val`s that needs to replay pad_op on. The map here
+  // stores the number of encounter for each `Val*` during traversal, the logic
+  // is used to decide whether propagation can cross Val with multiple uses in
+  // the graph. See NOTE [ Handling TV with Multiple Uses via Frontier ]
+  std::unordered_map<Val*, int64_t> frontier;
+  // replay_sequence is used later to create the updated branch with padded
+  // inputs after all `frontier` has been updated with padding.
+  std::vector<Expr*> replay_sequence;
```

Comment:
I suspect it's unnecessary to record a **sequence** and replay things in the sequence all at once. Instead, we may be able to simply move `pad` up incrementally, one "swap" (of `pad` and its preceding op) at a time.